File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Python/3SumClosest.py
"""
Given an array S of n integers, find three integers in S such that the sum is closest to a given number, target. 
Return the sum of the three integers. You may assume that each input would have exactly one solution.

For example, given array S = {-1 2 1 -4}, and target = 1.

that is closest to the target is 2. (-1 + 2 + 1 = 2).
"""
import sys

class Solution:
    # @return an integer
    def threeSumClosest(self, num, target):
        """
        Find the three number add up closest to target
        """
        if len(num) < 3:
            return -1   #error

        num = sorted(num)
        #print num
        curr_match = sys.maxint 
        result_sum = 0
        
        for idx in range( len(num)-2 ):
            left, right= idx + 1, len(num) -1

            while left < right:
                curr_sum = num[idx] + num[left] + num[right] 
                # Update curr_match and result_sum
                if abs(curr_sum - target) < curr_match:
                    curr_match = abs(curr_sum- target)
                    result_sum = curr_sum
                    print curr_match
                
                if curr_sum > target:
                    right = right - 1
                elif curr_sum < target:
                    left = left + 1
                else:
                    return target
        
        return result_sum

# Unit Test
if __name__ == "__main__":
    #num = [-1, 2, 1, -4]
    #target = 1
    #inst = Solution()
    #print inst.threeSumClosest(num, target)

    num = [1,1,1,0]
    target = -100
    inst = Solution()
    print inst.threeSumClosest(num, target)

File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/CPP/3SumClosest.cpp
/*

Given an array S of n integers, find three integers in S such that the sum is closest to a given number, target. Return the sum of the three integers. You may assume that each input would have exactly one solution.

    For example, given array S = {-1 2 1 -4}, and target = 1.

    The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
*/

class Solution {
public:
    int threeSumClosest(vector<int> &num, int target) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        int min = INT_MAX, sum, size = num.size();
        if( size < 3 )  return -1;    
        sort(num.begin(), num.end() );
        for( int i =0; i < size - 2; i ++ ){
            int left = i +1, right = size -1;
            
            while( left < right ){
                int temp_sum = num[i] + num[left] + num[right];
                int product = temp_sum - target;
                if( product < 0 ){
                    if( abs(product ) < min ){
                        min = abs( product );
                        sum = temp_sum;
                    }
                    left++;
                }
                else if( product > 0 ){
                    if( product< min ){
                        min = product;
                        sum = temp_sum;
                    }
                    right--;
                }
                else
                    return temp_sum;
            } 
        }
        return sum;
    }
};

int main(){
  Solution test;
  vector<int> num;
  int target;
  test.threeSumClosest(num, target);
}
